src/zkproof/mod.rs

```diff
@@ -0,0 +1,159 @@
+use super::*;
+use arithimpl::traits::*;
+
+use num_traits::{Zero, One};
+use std::ops::{Sub, Mul, Rem};
+
+use crypto::sha2::Sha256;
+use crypto::digest::Digest;
+
+use std::error::Error;
+use std::fmt;
+
+impl fmt::Display for ZKVerifierError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "ZKVerifierError")
+    }
+}
+
+impl Error for ZKVerifierError {
+    fn description(&self) -> &str {
+        "Error while verifying"
+    }
+}
+
+impl<I> ZKVerifier<I> for EncryptionKey<I>
+    where
+        I : Samplable,
+        I : Eq,
+        I : One,
+        I : Zero,
+        I : ModPow,
+        I : Sub<I, Output=I>,
+        I : EGCD,
+        I : ToString,
+        I : FromString<I>,
+        for<'a>        I: Rem<&'a I, Output=I>,
+        for<'a,'b> &'a I: Rem<&'b I, Output=I>,
+        for<'a>    &'a I: Mul<I, Output=I>,
+{
+    fn generate_challenge(&self) -> (Vec<I>, I, Vec<I>, Vec<I>) {
+        let (mut y, mut challenge) : (Vec<I>, Vec<I>) = (Vec::new(), Vec::new());
```

Comment:
could take advantage of the fact that we know the final size of the vectors in advance (unless simplified as mentioned below)